src/ascii/chain.rs

```diff
@@ -0,0 +1,386 @@
+        })
+    }
+}
+
+/// Information about a device.
+#[derive(Debug, Copy, Clone)]
+pub(crate) struct DeviceInfo {
+    /// The address of the device
+    pub address: NonZeroU8,
+    /// The number of axes
+    pub axis_count: u8,
+}
+
+impl DeviceInfo {
+    /// Collect a [`Vec`] of [`DeviceInfo`]s by inspecting the port.
+    pub fn collect_from_port<B: Backend>(
+        port: &mut Port<'_, B>,
+    ) -> Result<Vec<DeviceInfo>, AsciiError> {
+        let replies = port
+            .command_replies_until_timeout_with_check("get system.axiscount", check::flag_ok())?;
+        let mut unique_addresses = std::collections::HashSet::with_capacity(replies.len());
+        let mut infos = Vec::with_capacity(replies.len());
+        for reply in replies {
+            // Replies are always from non-zero addresses so it is safe to unwrap.
+            let address: NonZeroU8 = reply.target().device().try_into().unwrap();
+            if !unique_addresses.insert(address) {
+                return Err(DuplicateAddressError {
+                    address: address.get(),
+                }
+                .into());
+            }
+            let axis_count = reply.data().parse().map_err(|_| {
+                AsciiCheckDataError::new(
+                    "could not parse data as 8-bit unsigned integer",
+                    reply.into(),
+                )
+            })?;
+            infos.push(DeviceInfo {
+                address,
+                axis_count,
+            })
+        }
+        // To make future access more efficient sort the list of devices by address.
+        infos.sort_by_key(|info| info.address);
+        Ok(infos)
+    }
+}
+
+/// An iterator of the devices in a [`Chain`].
+#[derive(Debug)]
+pub struct ChainIter<'i, 'a, B, P> {
+    chain: &'i Chain<'a, B, P>,
+    next_device_address: NonZeroU8,
+}
+
+impl<'i, 'a, B, P: Clone> std::iter::Iterator for ChainIter<'i, 'a, B, P> {
+    type Item = Device<'a, B, P>;
+    fn next(&mut self) -> Option<Self::Item> {
+        self.chain
+            .device(self.next_device_address.get())
+            .map(|device| {
+                self.next_device_address = self.next_device_address.saturating_add(1);
+                device
+            })
+    }
+}
+
+/// A by-value iterator over the devices in a chain.
+#[derive(Debug)]
+pub struct ChainIntoIter<'a, B, P> {
+    chain: Chain<'a, B, P>,
+    next_device_address: NonZeroU8,
+}
+
+impl<'a, B, P: Clone> std::iter::Iterator for ChainIntoIter<'a, B, P> {
+    type Item = Device<'a, B, P>;
+    fn next(&mut self) -> Option<Self::Item> {
+        self.chain
+            .device(self.next_device_address.get())
+            .map(|device| {
+                self.next_device_address = self.next_device_address.saturating_add(1);
+                device
+            })
+    }
+}
+
+/// Represents a single device (controller or integrated product).
+#[derive(Debug, Clone)]
+pub struct Device<'a, B, P> {
+    port: P,
+    info: DeviceInfo,
```

Comment:
Is this the best structure? If a user sends a broadcast renumber, all devices in the chain could have incorrect metadata. Perhaps it would be better to have `SharedMut` references to their device info. That way if we implement a renumber command, we can update device metadata and all existing devices will be updated. 

However, this behavior could be confusing because address 1 may be a different device now. It certainly seems bad for them to continue talking to device at address A if it is not what we thought it was.

Alternatively, we could have existing `Device`s become invalid and no longer work (panic) if their addresses have changed. Panicking at runtime isn't ideal. Maybe we could force methods like renumber to fail if there are any shared devices out there? 

Think more about this.

Reply:
This has been refactored so that all chain information is now shared.